Adds spec for FileUploadTask (#27)
microsoftgraph/msgraph-sdk-design: tasks/FileUploadTask.md

# File Upload Task

## Objectives

Uploading a large file can be done by creating an upload session which allows the upload of chunks of the file in sequential API requests. This therefore enables the possibility mechanisms such as resuming the upload or cancelling the upload.
This task aims to provide a fluent and easy to use mechanism for the consumer to control, monitor and cancel the handling of large uploads.

## Requirements

> - Accept a Microsoft.Graph.UploadSession instance to be used for the upload which is created and managed by the user and handed over to the task. The task should use the instance to start/resume the upload.

Reviewer: We should validate the expiration and byte range on resume by [querying the upload URL](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0#resuming-an-in-progress-upload).

Author: I've added this to the spec

- Accept a file/stream to be uploaded
- The task should provide a mechanism of configuring the upload properties i.e. the upload chunk size
- The task should provide a mechanism of monitoring the upload status i.e the progress and failures through a callback/Delegate
- Using the RequestContext, the flag for the FileUploadTask can be set.
- In the event that a chunk fails to upload, the task should retry to re-upload the chunk for a configurable number of times(default of 3) before giving up and throwing an error/exception for the user to handle..

## Performance Considerations

- The FileUploadTask should not unnecessarily create new streams/buffers to read or write to for the purposes of the small chunks. The memory consumed should be as minimal as possible.

## Telemetry considerations

- Using the RequestContext, the flag for the FileUploadTask can be set to enable monitoring of the use of the task.

## Example usage

### CSharp

Create a handler for the progress of the upload

```CSharp

> public class ProgressCallback : IProgressCallback

Reviewer: We should consider using the [AsyncMonitor](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/blob/dc8ae2e7cc49a8f2240d2cba707ef52044ecdfa6/src/Microsoft.Graph.Core/Requests/AsyncMonitor.cs) for this. I don't think we use anywhere so it is worth a look.

Author: To me, it looks like the AsyncMonitor is better suited for a long running operation running on graph service (e.g. [creating a copy](https://docs.microsoft.com/en-us/graph/long-running-actions-overview?tabs=http) of a large file already on onedrive) and one would be polling the service for the progress at a frequency of their choosing. 
However, we want to provide a feed back mechanism to what the client SDK is doing for the user and provide mechanisms to show progress and react to failures as the client SDK does the upload task.
@MIchaelMainer

Reviewer: Makes sense.

# File Upload Task

## Objectives

> Uploading a large file can be done by creating an upload session which allows the upload of chunks of the file in sequential API requests. This therefore enables the possibility mechanisms such as resuming the upload or cancelling the upload.

Reviewer 2: ```suggestion
Uploading a large file can be done by creating an upload session which allows the upload of slices of the file in sequential API requests. This therefore enables the possibility mechanisms such as resuming the upload or cancelling the upload.
```

This task aims to provide a fluent and easy to use mechanism for the consumer to control, monitor and cancel the handling of large uploads.

## Requirements

- Accept a Microsoft.Graph.UploadSession instance to be used for the upload which is created and managed by the user and handed over to the task. The task should use the instance to start/resume the upload.
- The task should validate the expiration time and expected byte range of the upload session on resume by querying the upload URL(thereby updating its instance of the Microsoft.Graph.UploadSession).
- Accept a file/stream to be uploaded

> - The task should provide a mechanism of configuring the upload properties i.e. the upload chunk size

Reviewer 2: ```suggestion
- The task should provide a mechanism of configuring the upload properties i.e. the upload slice size
```

- The task should provide a mechanism of monitoring the upload status i.e the progress and failures through a callback/Delegate

> - Using the RequestContext, the flag for the FileUploadTask can be set.

Reviewer 2: ```suggestion
- Using the RequestContext, the feature flag for the FileUploadTask can be set for telemetry purposes.
```

- In the event that a chunk fails to upload, the task should retry to re-upload the chunk for a configurable number of times(default of 3) before giving up and throwing an error/exception for the user to handle.

## Performance Considerations

> - The FileUploadTask should not unnecessarily create new streams/buffers to read or write to for the purposes of the small chunks. The memory consumed should be as minimal as possible.

Reviewer 2: ```suggestion
- The FileUploadTask should not create buffers for more than the current slice of data being uploaded.
```

- Using the RequestContext, the flag for the FileUploadTask can be set to enable monitoring of the use of the task.

## Example usage

### CSharp

Create a handler for the progress of the upload

```CSharp

public class ProgressCallback : IProgressCallback
{
    public void UpdateProgress(long current, long max)
    {
        //Check progress  
        Console.WriteLine("Processed " + current + "of" + max);
    }

    public void OnSuccess( DriveItem result )
    {
        //Handle the successful response
        Console.WriteLine("Upload complete");
    }

    public void OnFailure( ClientException ex )
    {
        //Handle the failed upload
        Console.WriteLine("Upload failed");
    }
}
```

Create an upload session using the request builders

```CSharp
// create an upload session
var uploadSession = await graphClient.Drive.Items["01KGPRHTV6Y2GOVW7725BZO354PWSELRRZ"].ItemWithPath("_hamilton.png").CreateUploadSession().Request().PostAsync();
```

Use the upload session and callback handler to run/resume the upload

```CSharp
// create the upload provider with relevant options
var maxChunkSize = 320 * 1024; // 320 KB - Change this to your chunk size. 5MB is the default.

> var provider = new ChunkedUploadProvider(uploadSession, graphClient, stream, maxChunkSize);

Reviewer 2: As we have already released ChunkUploadProvider, we need to keep this in C#, however, we should wrap this in a LargeFileUploadTask, to fix the naming.  I have no idea how we ended up copying the C# design in Java.  We were supposed to follow the JavaScript design.

Author: This make sense to me. 
I have updated the sample in the spec to reflect this for C#.
